Add filter support

joshshadowfax opened the pull request:
Support filters provided by the API to allow searches by more than just name.  Also fixed issue where only the first 100 results were being randomly chosen from due to API pagination (now card will be randomly grabbed from all results even if there are more than 100).

joshshadowfax committed:
Add filter support

joshshadowfax commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
-    url = url.format(quotedsearchterm)
-    return magic(url)
+#find number of results and get a random result
+def getresult(url):
```
API paginates with 100-page results.  There's not a function to get total number of results for a query according to the API, so we're binary searching to find the number of results, then picking a random result from all those.  This shouldn't be too intensive (total number of magic cards is <14000 so shouldn't run into overflow issues or too many searches).

kesre commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
-    url = "https://api.deckbrew.com/mtg/cards?name={}"
-    url = url.format(quotedsearchterm)
-    results = requests.get(url).json()
+filterlist = ["type","supertype","subtype","name","oracle","set","rarity","set","color","multicolor","m","format","status"]
```
let's multiline the filter list.

kesre commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
+def search(searchterm):
+    filters = getfilters(searchterm)
+    if filters:
+        search = reduce(lambda string, term: string + ("" if not string else "&") + "{0}={1}".format(term[0].strip(),quote(term[1].strip())), filters, "")
```
This needs to be broken up some. Still trying to think of a good way to do it.
I'd put this in a list comp + join rather than a reduce

See below comment on getfilters

kesre commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
+filterlist = ["type","supertype","subtype","name","oracle","set","rarity","set","color","multicolor","m","format","status"]
 
-    if not results:
+def search(searchterm):
```
Looks like we quote() searchterm a few times. Can we just do it once at the beginning?

joshshadowfax replied:
Quote is only used twice, once with searchterm and once in the reduce.  Not sure where you're seeing other uses.

kesre replied:
Oh. forgot to comment the searchterm.lower() below. Seems like we'd want to make sure we quote everywhere.

joshshadowfax replied:
Checking for exact match requires the unquoted searchterm, since the API returns card names in an unquoted format.

kesre commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
+    else:
+        #if no filters found we default to name search
+        url = "https://api.deckbrew.com/mtg/cards?name={}".format(quote(searchterm))
+        results = requests.get(url).json()
```
is it possible this is paginated as well?

joshshadowfax replied:
It could be paginated.  Will need a slight rework to check all pages for exact match.

kesre commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
+            lo = page
+        else:
+            hi = page - 1
+    results = 100*(lo-1) + len(requests.get(url + "&page={}".format(lo-1)).json())
```
Let's make the 100 a named constant - or could we get this from the number of items returned on the first page?

joshshadowfax replied:
We can get this from the number of items on the first page, in case the number per page changes.

kesre commented on plugins/magic.py:
```diff
@@ -1,39 +1,67 @@
+    choice = random.randint(1,results) - 1
+    return requests.get(url + "&page={}".format(choice/100)).json()[choice%100]
 
+def getfilters(searchterm):
```
Hmm might be cool to implement getfilters in such a way that +general expands to / overwrites the applicable filters. Maybe make the filters a dict so we retain which filter it's used for? That could also make collapsing the filters to a url a little more straightforward (some iteritems call).

"&".join(["{0}={1}".format(param, value) for param, value in filters.iteritems()])

joshshadowfax committed:
Updated magic:
- Now uses Gatherer images
- Filter refactor to better support future filter types
- Pagination handling

Also added amazing, a plugin which will return an image of the LoL pro
player Amazing.
